### PR TITLE
Move ledger methods into Ledger struct

### PR DESCRIPTION
**Pull request by sisuresh**
Addresses PR comments from https://github.com/stellar/rs-soroban-sdk/pull/455.

### CODE REVIEW

File: soroban-sdk/src/ledger.rs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        &self.0
+    }
+
+    #[inline(always)]
+    pub(crate) fn new(env: &Env) -> Ledger {
+        Ledger(env.clone())
+    }
+
+    /// returns the current ledger version
+    pub fn get_ledger_version(&self) -> u32 {
```
Comment (leighmcculloch) [inline soroban-sdk/src/ledger.rs:21]: Let's follow the patterns of other docs in the SDK, by writing sentences. For example:
https://github.com/stellar/rs-soroban-sdk/blob/bba4039966263f609ceefbf5b30a43361eecaa01/soroban-sdk/src/contract_data.rs#L79-L90

I think we should skip repeating `ledger` in the function name given it is a function on the `Ledger` type.

I think we should add the term protocol too. This isn't the version of the ledger, it's the version of the protocol that this ledger was created with. 

```suggestion
    /// Returns the version of the protocol that the ledger created with.
    pub fn protocol_version(&self) -> u32 {
```

If any of these functions panic we should add a `### Panics` section as well, which is the convention. As far as I can tell these functions will never panic, but if they do can you add sections stating when they do?

**Review by leighmcculloch — COMMENTED** [review]
Thanks for adding the comments!

Some requests inline.

File: soroban-sdk/src/ledger.rs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+    /// returns a unix timestamp of when the netowrk closed the last ledger
+    pub fn get_ledger_timestamp(&self) -> u64 {
+        let env = self.env();
+        let obj = internal::Env::get_ledger_timestamp(env);
+        internal::Env::obj_to_u64(env, obj)
+    }
+
+    /// returns the network passphrase
+    pub fn get_ledger_network_id(&self) -> Bytes {
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply soroban-sdk/src/ledger.rs]: That's a good question. The host calls it `network_id`, so this change mirrors that. @jayz22 any objection to calling it `network_passphrase` in the sdk?

File: soroban-sdk/src/ledger.rs (hunk 1 above)
Comment (sisuresh) [reply soroban-sdk/src/ledger.rs]: It doesn't look like these functions panic.

File: soroban-sdk/src/ledger.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jayz22) [reply soroban-sdk/src/ledger.rs]: Yeah looks like "network passphrase" is the correct term. No objection.

[the remark this answers is not in the document]
Comment (sisuresh) [reply soroban-sdk/src/ledger.rs]: Done

File: soroban-sdk/src/ledger.rs [hunk 3]
```diff
@@ -0,0 +1,47 @@
+use crate::{
+    env::internal::{self, RawValConvertible},
+    Bytes, Env,
+};
+
+#[derive(Clone)]
+pub struct Ledger(Env);
```
[the remark this answers is not in the document]
Comment (sisuresh) [reply soroban-sdk/src/ledger.rs:30]: Done

File: soroban-sdk/src/ledger.rs [hunk 4]
```diff
@@ -0,0 +1,83 @@
+use crate::{
+    env::internal::{self, RawValConvertible},
+    Bytes, Env,
+};
+
+/// Ledger allows retrieval of ledger information
```
Comment (leighmcculloch) [inline soroban-sdk/src/ledger.rs:6]:
```suggestion
/// Ledger allows retrieval of ledger information.
```

File: soroban-sdk/src/env.rs [hunk 5]
```diff
@@ -105,6 +105,12 @@ impl Env {
         ContractData::new(self)
     }
 
+    /// Get a [Ledger] for accessing ledger information
```
Comment (leighmcculloch) [inline soroban-sdk/src/env.rs:108]:
```suggestion
    /// Get a [Ledger] for accessing the current ledger.
```

File: soroban-sdk/src/env.rs [hunk 6]
```diff
@@ -244,30 +250,9 @@ impl Env {
...
-    }
-
-    #[doc(hidden)]
-    pub fn get_ledger_network_id(&self) -> Bytes {
-        let bin_obj = internal::Env::get_ledger_network_id(self);
-        bin_obj.in_env(self).try_into().unwrap()
-    }
 }
 
+use crate::ledger::Ledger;
```
Comment (leighmcculloch) [inline soroban-sdk/src/env.rs:255]: Could this be inserted at the top of the file with the other imports? The imports where it has been grouped are testutils imports.

**Review by leighmcculloch — APPROVED** [review]

File: soroban-sdk/src/env.rs [hunk 7]
```diff
@@ -296,9 +283,27 @@ impl Env {
...
+            network_id: vec![0u8],
+        };
+        env_impl.set_ledger_info(l);
+        Env { env_impl }
+    }
+
+    /// Creates an [Env] with empty storage and a user specified [LedgerInfo]
+    pub fn with_empty_recording_storage_and_ledger_info(li: LedgerInfo) -> Env {
+        let env_impl = Self::with_empty_recording_storage_env_impl();
+        env_impl.set_ledger_info(li);
```
Comment (leighmcculloch) [inline soroban-sdk/src/env.rs:305]: The layering of these functions is not going to scale well to us setting other attributes on `Env` in tests.

Can we add a testutils function `set_ledger` (drop the `_info`) to `Env` instead? So the idea being someone will do this in a test:
```rust
let env = Env::default();
env.set_ledger(LedgerInfo{ ... });
```

Comment (sisuresh) [reply soroban-sdk/src/env.rs]: Done

File: Cargo.toml [hunk 8]
```diff
@@ -12,7 +12,6 @@ members = [
     "tests/import_contract",
     "tests/invoke_contract",
     "tests/udt",
-    "tests/contract_data",
```
Comment (leighmcculloch) [inline Cargo.toml:15]: Why is the contract data example removed?

Comment (leighmcculloch) [reply Cargo.toml:15]: I re-added in b1fcf44.

**Review by leighmcculloch — APPROVED** [review]
🎉 Thank yoU!

File: soroban-sdk/src/env.rs [hunk 9]
```diff
@@ -244,33 +249,13 @@ impl Env {
...
-    pub fn get_ledger_network_id(&self) -> Bytes {
-        let bin_obj = internal::Env::get_ledger_network_id(self);
-        bin_obj.in_env(self).try_into().unwrap()
-    }
 }
 
 #[cfg(feature = "testutils")]
 use crate::testutils::ContractFunctionSet;
 #[cfg(feature = "testutils")]
+use soroban_env_host::LedgerInfo;
```
Comment (leighmcculloch) [inline soroban-sdk/src/env.rs:257]: 🤔 It doesn't look like LedgerInfo is exported, so I don't think contracts can actually call `set_ledger` with it.

Comment (leighmcculloch) [reply soroban-sdk/src/env.rs]: Exported in 66797858c05123e6f426b20a093403dc5dd77372.

File: Cargo.toml (hunk 8 above)
Comment (sisuresh) [reply Cargo.toml]: Ah yeah this was a mistake. Thanks for fixing it.

**Pull request merged**